### ISSUE TITLE
Make VoiceAssistant support multi Participants

### ISSUE THREAD

**Issue by Nicj228** [livekit/agents#391]
Feat ;  Actualy the voice is only start for one participant ! so to forwad transcription  and  use other feat in the voice assistant for each participant  we have to start many VoiceAssistant   or add custom Transcription !

**Comment by theomonnom** [comment]
Hey, we plan to support multiple participants for the same Voice Assistant in the future, but this isn't a priority at the moment.

**Comment by theomonnom** [comment]
We greatly appreciate contributions!

**Comment by Shandelier** [comment]
How's the work on the multiple participants support? That would help greatly!
 
One of the use cases would be a live translation via agent. Most of the logic is already there in voice assistant

**Comment by longcw** [comment]
The switching between multiple participants will be supported in agent 1.0, so there could be multiple participants in the room and there will be an API to select which one to talk to the agent, all the participants can hear from the agent. We should have a beta release very soon.

**Comment by RobMaye** [comment]
> The switching between multiple participants will be supported in agent 1.0, so there could be multiple participants in the room and there will be an API to select which one to talk to the agent, all the participants can hear from the agent. We should have a beta release very soon.

Great! Thanks, appreciate the update :)

**Comment by nitishymtpl** [comment]
Hey@longcw, Any updates or corresponding examples?

**Comment by longcw** [comment]
@nitishymtpl Yes, we just released the agents 1.0 rc, there is an example about how to switch participants https://github.com/livekit/agents/blob/dev-1.0/examples/voice_agents/toggle_io.py#L14

Check here https://github.com/livekit/agents?tab=readme-ov-file#-10-release- to see how to install the rc version.

**Comment by longcw** [comment]
You may create multiple `RoomIO` instances, only one of them has the output enabled and each connects audio input to a participant, but you need then to handle the multiple audio inputs, like mixing them or sending them to different STT?

**Comment by nitishymtpl** [comment]
Hi @longcw, Only problem is AI agent could listen to only one participant. Since AI agent is sending the output into room, both participants could hear the response of AI Agent. Somehow if we could subscribe the microphone of both participants then AI agent could hear from both of them. Currently it is subscribed to only one microphone. I’m curious if you could check this. Since you are expert on this, you could easily fix it :). Pls do share repo or corresponding example.

**Comment by longcw** [comment]
This could be an interesting use case to have multi user in the room.

Reading audio streams and mixing them from multiple participants is feasible, but the problem is how the AI agent to hear from both, is that a realtime API or a STT model takes the audio?  You will need the model to support distinguishing the speakers.

**Comment by nitishymtpl** [comment]
We could use any one of them: realtime API or STT pipeline. There could two possibilities: 
1. Subscribing to both microphones- mixing the audio streams and passing into AI Agent. AI agent don’t need to distinguish speaker. Based on the input, AI agent would simply translate and send the audio into room so that both participants could hear. E.g: if someone is speaking French, it would convert into English and if someone is speaking English it would convert into French. They don’t need to distinguish speaker.
2. Subscribing to both microphones- but having separate channels for both participants. Audio streams from participants would be going to AI Agent separately. Since AI agent is simply translating into other languages. It would serve the purpose. Let me know if it’s feasible.

**Comment by OrkhanGG** [comment]
> This could be an interesting use case to have multi user in the room.
> 
> Reading audio streams and mixing them from multiple participants is feasible, but the problem is how the AI agent to hear from both, is that a realtime API or a STT model takes the audio? You will need the model to support distinguishing the speakers.

I don’t think any publicly available LLM API currently supports multi-participant audio natively, and I doubt we’ll see that change soon. However, here are a few possible approaches:

1. The LLM or speech-to-text model supports speaker diarization (best case). This allows it to distinguish between different speakers automatically.

2. The model accepts metadata along with the audio input. In this case, you could manually label each speaker and prepend their identity before merging the audio. This would require some preprocessing.

3. A complex and hacky workaround (not recommended for anything beyond experimental stages): you could prepend each participant’s audio with a short TTS-generated snippet like “Username: {username}” and cache these in memory. Then, when a participant speaks, you detect the activity, insert their identifier snippet before their actual audio, and send the whole thing to the LLM. For example, the final audio sent might sound like: “Username: Alice. Hey, can you find my data?” This gives the LLM a chance to infer who is speaking. But again, this approach is messy and error-prone, and should really only be used in prototypes.

**Comment by Khalid1G** [comment]
Hey @OrkhanGG, @theomonnom  , Any updates or corresponding examples?

**Comment by panosmoschos** [comment]
Hello guys do we have any update regarding this? It would be extremely useful if you could provide an example of how we can have one agwnt that can support conversations with multiple participants!

**Comment by OrkhanGG** [comment]
> Hey [@OrkhanGG](https://github.com/OrkhanGG), [@theomonnom](https://github.com/theomonnom) , Any updates or corresponding examples?

I initially achieved this using some very hacky methods by building my own agent implementation from scratch with LiveKit. However, with the new Gemini Live API model, the process is now much easier and more robust. I'm still in the process of refactoring the code, but if you try the "Gemini 2.5 Flash Preview Native Audio Dialog", things get even simpler (you can provide model name as 'gemini-2.5-flash-preview-native-audio-dialog'). It supports proactive dialogue and several additional parameters that can be useful for multi-participant audio rooms.

You can check out this example to get started:
https://github.com/livekit/agents/blob/main/examples/voice_agents/push_to_talk.py

There are also some advanced possibilities you might explore, like detecting the active speaker and having the agent listen only to them. Supporting simultaneous audio and video input from all participants would be a huge win, though I don't think current LLMs can easily handle that out of the box. Any alternative solution would likely require enterprise-level effort to bring into production.

**Comment by Nicj228** [comment]
Good job. I'll test it

Le mer. 2 juil. 2025, 22:40, Orkhan Aliyev ***@***.***> a
écrit :

**Comment by panosmoschos** [comment]
> > Hey [@OrkhanGG](https://github.com/OrkhanGG), [@theomonnom](https://github.com/theomonnom) , Any updates or corresponding examples?
> 
> I initially achieved this using some very hacky methods by building my own agent implementation from scratch with LiveKit. However, with the new Gemini Live API model, the process is now much easier and more robust. I'm still in the process of refactoring the code, but if you try the "Gemini 2.5 Flash Preview Native Audio Dialog", things get even simpler (you can provide model name as 'gemini-2.5-flash-preview-native-audio-dialog'). It supports proactive dialogue and several additional parameters that can be useful for multi-participant audio rooms.
> 
> You can check out this example to get started: https://github.com/livekit/agents/blob/main/examples/voice_agents/push_to_talk.py
> 
> There are also some advanced possibilities you might explore, like detecting the active speaker and having the agent listen only to them. Supporting simultaneous audio and video input from all participants would be a huge win, though I don't think current LLMs can easily handle that out of the box. Any alternative solution would likely require enterprise-level effort to bring into production.


Is the code in the example somehow relatable to which model you are using? I mean if it works on Gemini can i make it work with openai with some fine tune or not? I am asking because currently our whole infra is on openai!

**Comment by OrkhanGG** [comment]
> > > Hey [@OrkhanGG](https://github.com/OrkhanGG), [@theomonnom](https://github.com/theomonnom) , Any updates or corresponding examples?
> > 
> > 
> > I initially achieved this using some very hacky methods by building my own agent implementation from scratch with LiveKit. However, with the new Gemini Live API model, the process is now much easier and more robust. I'm still in the process of refactoring the code, but if you try the "Gemini 2.5 Flash Preview Native Audio Dialog", things get even simpler (you can provide model name as 'gemini-2.5-flash-preview-native-audio-dialog'). It supports proactive dialogue and several additional parameters that can be useful for multi-participant audio rooms.
> > You can check out this example to get started: https://github.com/livekit/agents/blob/main/examples/voice_agents/push_to_talk.py
> > There are also some advanced possibilities you might explore, like detecting the active speaker and having the agent listen only to them. Supporting simultaneous audio and video input from all participants would be a huge win, though I don't think current LLMs can easily handle that out of the box. Any alternative solution would likely require enterprise-level effort to bring into production.
> 
> Is the code in the example somehow relatable to which model you are using? I mean if it works on Gemini can i make it work with openai with some fine tune or not? I am asking because currently our whole infra is on openai!

Yes, you can definitely do that. The example I shared above demonstrates how to set up an agent that focuses on a specific participant. The trick is to create the illusion that the agent is listening to all participants.

If you don’t want the agent to focus on a specific participant based on client-side events, you can use something like active speaker detection instead. This might require a few extra steps to implement properly, but in my opinion, it's definitely worth considering and is very likely to work well with OpenAI's real-time voice API.

**Comment by Khalid1G** [comment]
> > > > Hey [@OrkhanGG](https://github.com/OrkhanGG), [@theomonnom](https://github.com/theomonnom) , Any updates or corresponding examples?
> > > 
> > > 
> > > I initially achieved this using some very hacky methods by building my own agent implementation from scratch with LiveKit. However, with the new Gemini Live API model, the process is now much easier and more robust. I'm still in the process of refactoring the code, but if you try the "Gemini 2.5 Flash Preview Native Audio Dialog", things get even simpler (you can provide model name as 'gemini-2.5-flash-preview-native-audio-dialog'). It supports proactive dialogue and several additional parameters that can be useful for multi-participant audio rooms.
> > > You can check out this example to get started: https://github.com/livekit/agents/blob/main/examples/voice_agents/push_to_talk.py
> > > There are also some advanced possibilities you might explore, like detecting the active speaker and having the agent listen only to them. Supporting simultaneous audio and video input from all participants would be a huge win, though I don't think current LLMs can easily handle that out of the box. Any alternative solution would likely require enterprise-level effort to bring into production.
> > 
> > 
> > Is the code in the example somehow relatable to which model you are using? I mean if it works on Gemini can i make it work with openai with some fine tune or not? I am asking because currently our whole infra is on openai!
> 
> Yes, you can definitely do that. The example I shared above demonstrates how to set up an agent that focuses on a specific participant. The trick is to create the illusion that the agent is listening to all participants.
> 
> If you don’t want the agent to focus on a specific participant based on client-side events, you can use something like active speaker detection instead. This might require a few extra steps to implement properly, but in my opinion, it's definitely worth considering and is very likely to work well with OpenAI's real-time voice API.

@OrkhanGG How is this going to work with the Gemini Live API, given that it includes built-in VAD-based turn detection, which is currently the only supported method? Is there any way to make it manual? can you provide an example.

**Comment by OrkhanGG** [comment]
> > > > > Hey [@OrkhanGG](https://github.com/OrkhanGG), [@theomonnom](https://github.com/theomonnom) , Any updates or corresponding examples?
> > > > 
> > > > 
> > > > I initially achieved this using some very hacky methods by building my own agent implementation from scratch with LiveKit. However, with the new Gemini Live API model, the process is now much easier and more robust. I'm still in the process of refactoring the code, but if you try the "Gemini 2.5 Flash Preview Native Audio Dialog", things get even simpler (you can provide model name as 'gemini-2.5-flash-preview-native-audio-dialog'). It supports proactive dialogue and several additional parameters that can be useful for multi-participant audio rooms.
> > > > You can check out this example to get started: https://github.com/livekit/agents/blob/main/examples/voice_agents/push_to_talk.py
> > > > There are also some advanced possibilities you might explore, like detecting the active speaker and having the agent listen only to them. Supporting simultaneous audio and video input from all participants would be a huge win, though I don't think current LLMs can easily handle that out of the box. Any alternative solution would likely require enterprise-level effort to bring into production.
> > > 
> > > 
> > > Is the code in the example somehow relatable to which model you are using? I mean if it works on Gemini can i make it work with openai with some fine tune or not? I am asking because currently our whole infra is on openai!
> > 
> > 
> > Yes, you can definitely do that. The example I shared above demonstrates how to set up an agent that focuses on a specific participant. The trick is to create the illusion that the agent is listening to all participants.
> > If you don’t want the agent to focus on a specific participant based on client-side events, you can use something like active speaker detection instead. This might require a few extra steps to implement properly, but in my opinion, it's definitely worth considering and is very likely to work well with OpenAI's real-time voice API.
> 
> [@OrkhanGG](https://github.com/OrkhanGG) How is this going to work with the Gemini Live API, given that it includes built-in VAD-based turn detection, which is currently the only supported method? Is there any way to make it manual? can you provide an example.

The project I’m working on is private, but I’ll put together a public demo soon and share it here unless the LiveKit team beats me to it! Hopefully, it’ll clear up your main concern about the “two drivers, one steering wheel” (two VADs and one LLM) setup.

**Comment by MatheusRDG** [comment]
Will be amazing have something like this, @OrkhanGG let us know if that works for you.

**Comment by deemedai** [comment]
any update on multiparticipant support or any useful workarounds  to achieve this

**Comment by KidesLeo** [comment]
Hey everyone, for those still looking for a workaround to capture and process audio from specific participants (like  a second user in the room), I managed to get this working by creating a standalone `rtc.AudioStream` from the target participant's track and feeding it directly into a separate STT stream.

As @longcw and @OrkhanGG discussed above regarding reading and mixing audio streams, you can isolate a specific participant's track and transcribe it independently.

Here is a simplified version of the approach I used:

```python
import asyncio
from livekit import rtc
from livekit.agents.stt import SpeechEventType

async def start_participant_transcription(participant_identity: str, room: rtc.Room, stt):
    participant = room.remote_participants.get(participant_identity)
    if not participant:
        return
        
    audio_track = None
    for pub in participant.track_publications.values():
        if pub.kind == rtc.TrackKind.KIND_AUDIO and pub.track:
            audio_track = pub.track
            break

    if not audio_track:
        return

    audio_stream = rtc.AudioStream.from_track(
        track=audio_track,
        sample_rate=16000,
        num_channels=1,
    )
    
    stt_stream = stt.stream()

    async def _transcribe():
        try:
            async def _feed_frames():
                async for event in audio_stream:
                    stt_stream.push_frame(event.frame)
                stt_stream.end_input()

            async def _read_transcripts():
                async for event in stt_stream:
                    if event.type != SpeechEventType.FINAL_TRANSCRIPT:
                        continue
                        
                    text = event.alternatives[0].text if event.alternatives else ""
                    if text.strip():
                        # Handle the transcript (e.g., inject into agent history)
                        print(f"[{participant_identity}]: {text}")

            await asyncio.gather(_feed_frames(), _read_transcripts())
        finally:
            await audio_stream.aclose()
            await stt_stream.aclose()

    return asyncio.create_task(_transcribe())
```

This works perfectly for my use case, but it could easily be adapted to listen to multiple participants by spawning a task like this for each user you want to track. 

I also ended up extending the transcript role so I could track which participant is being transcribed and append to the session history.

Hopefully, this helps anyone trying to bridge the gap before native multi-participant support is fully rolled out
